### PR TITLE
Add extended glib_wrapper syntax for cross-crate inheritance

### PR DESCRIPTION
**Pull request by gkoz**

### CODE REVIEW

File: src/object.rs [hunk 1]
```diff
@@ -232,9 +232,73 @@ macro_rules! glib_object_wrapper {
...
+
+        impl $crate::object::IsA<$super_name> for $name { }
+    };
+
+    (@munch_impls $name:ident, $super_name:path, $($implements:tt)*) => {
+        glib_object_wrapper!(@munch_impls $name, $super_name);
+        glib_object_wrapper!(@munch_impls $name, $($implements)*);
+    };
+
+    (@munch_impls $name:ident, $super_name:path => $super_ffi:path, $($implements:tt)*) => {
```
Comment (EPashkin) [inline src/object.rs]: Not safer place it before L288?

Comment (gkoz) [inline src/object.rs]: I don't expect any ambiguity: in L288 `$super_name` is followed by `,` and here by `=>`.

Comment (EPashkin) [inline src/object.rs]: Missed that.

File: src/wrapper.rs [hunk 2]
```diff
@@ -123,26 +139,25 @@ macro_rules! glib_wrapper {
...
     ) => {
-        glib_shared_wrapper!([$($attr)*] $name, $ffi_name, @ref $ref_arg $ref_expr,
-            @unref $unref_arg $unref_expr);
+        glib_object_wrapper!([$($attr)*] $name, $ffi_name, @get_type $get_type_expr, []);
     };
 
     (
         $(#[$attr:meta])*
-        pub struct $name:ident(Object<$ffi_name:path>);
+        pub struct $name:ident(Object<$ffi_name:path>): [$($implements:tt)+];
```
Comment (EPashkin) [inline src/wrapper.rs]: [] really needed?
May be change to something like:
```
pub struct Application(Object<ffi::GtkApplication>): 
         gio::Application(gio_ffi::GApplication),
         gio::ActionGroup(gio_ffi::GActionGroup),
         gio::ActionMap(gio_ffi::GActionMap);
```

Comment (gkoz) [inline src/wrapper.rs]: `[]` lets me capture its insides as a set of `tt`, that can be processed later. It's difficult if not impossible to work with a variable number of tokens somewhere other than at the right side of a token tree. `[]` provides a separate tree, without `[]` I'd get ambiguity with `;`.

So, without `[]` I'd have to make each entry use the extended syntax. Maybe that's not a big downside. Also note that `path` must be followed by either of `=>` `,` `=` `|` `;` `:` `>` `[` `{` `as` `where`.

Comment (EPashkin) [inline src/wrapper.rs]: Thank for explanation.

Comment (EPashkin) [inline src/wrapper.rs]: If I not wrong again then  `pub struct Bin(Object<ffi::GtkBin>): Container, Widget, Buildable;` not usable now and
example in need be fixed also.

Comment (gkoz) [inline src/wrapper.rs]: Since travis is green I assume the old syntax keeps working.

Comment (gkoz) [inline src/wrapper.rs]: I could add a case to gir so it wouldn't generate `[]` if they aren't needed.

**Comment by EPashkin** [comment]
Thanks.

**Comment by EPashkin** [comment]
Merge then.

**Comment by GuillaumeGomez** [comment]
Good for me as well. Thanks @gkoz!

**Pull request merged**